File: ml-benchmark/synduce/tree/maxtree.f
Inductive Tree = nil Unit | node {Int, Tree, Tree};

max = \a: Int. \b: Int. if < a b then b else a;

spec = (fix (
  \f: Int -> Tree -> Int. \w: Int. \t: Tree.
  match t with
    nil _ -> w
  | node {a, l, r} -> f (max (f w l) a) r
  end
)) 0;

repr = fix (
  \f: Tree -> Compress Tree. \t: Tree.
  match t with
    nil _ -> nil unit
  | node {a, l, r} -> node {a, f l, f r}
  end
);

main = \t: Tree. spec (repr t);

File: ml-benchmark/synduce/tree/maxtree.ml
type tree = Nil of unit | Node of int * tree * tree

let max a b = if a < b then b else a

let rec f w t =
  match t with
  | Nil _ -> w
  | Node (a, l, r) -> f (max (f w l) a) r

let spec = f 0

let rec repr t =
  match t with
  | Nil _ -> Nil ()
  | Node (a, l, r) -> Node (a, repr l, repr r)

let program t = spec (repr t)